### ISSUE TITLE
Add option to always open files at their canonical path

### ISSUE THREAD

**Issue by koenlek** [microsoft/vscode#130082]
# Problem

Symlinks can sometimes lead to the same files being opened more than once in VSCode. An example in which this happens is in a `bazel run` sandbox, which symlinks files together into a sandbox environment. Another example I've seen is when python modules (spread over various folders in a git repo) are symlinked into 1 virtual env python path tree. More generally speaking, symlinks are sometimes used by build tooling to create one consolidated environment out of files spread over different places. 

Unfortunately, this can lead to a poor user experience in VS Code. Debuggers can start opening multiple copies of the same file (the one you'd like to edit, which likely is under source control, and one at the path at which it appears in the sandbox) and the UI/UX experience is clumsy at best (breakpoints can be set in each copy of the file and actually work for both, but visually aren't synced). Language servers like pylance can resolve imports to the symlinked venv. So opening the definition of a function can open the symlinked copy, rather than the original (which is likely under source control and the one that you'd actually like to edit)... Again: clumsy UI/UX, such as that that symlinked copy doesn't show source control change info (lines added/rm/changed, gitlens).

# Proposed solution

I've seen various tickets related to this and generally things seem to get stuck in the discussion on defining a solution. I'd like to take another attempt at that. 

_How about we introduce a boolean option `openFilesAtCanonicalPath` (which people can set on a user or a workspace level)?_ 

The "canonical path" would be the path that `readlink -f {some_path}` would give on linux: i.e. the path with all symlinks resolved. This is the 'original' path for the file. This would:
1. Make sure there's always only 1 copy of a file opened
2. Have the file opened at the path that the user is most likely interested in. If it's part of the workspace, then this would be the path that is under version control (and all things like git/gitlens would work). If it's not part of the workspace, so be it, it's likely some system level dependency or so, still you'd be looking at the original file, which is great).

I wonder what VS Code devs thinks of this. 

Related tickets: https://github.com/microsoft/pylance-release/issues/478, https://github.com/microsoft/vscode/issues/34627, https://github.com/microsoft/vscode/issues/100533, https://github.com/microsoft/vscode-python/discussions/15897

**Final remarks**

We could even expand this idea to extensions if desired (extensions could retrieve both the requested path and canonical path of a file through extensions api), though I think many of the discussions around better extension support for symlinked files are maybe already solved by just opening the canonical file instead.

Please note that I'm focussing purely on this problem as introduced by symlinks in macOS and Linux. I'm not sure about Windows or other types of links like hard links, macOS aliases, etc. So far, I've seen build tooling use symlinks for these kind of things so addressing that can hopefully solve the majority of issues out there.

**Comment by GeekyMonkey** [comment]
Getting this onto the Backlog milestone is not much to celebrate. There are currently over 2900 tickets in that milestone.  How do we get this bumped to a milestone that will actually happen?

**Comment by misha-antonenko** [comment]
Hello, how would one approach fixing this? Is this time-consuming?

One of the problems not mentioned is, if the symbolic link is situated in a `bazel-bin` directory or a like one, VSCode won't correctly show git blames for now. It only works in a concurrently open tab with the original file

**Comment by ludakhris** [comment]
Any updates?

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!

**Comment by avkonst** [comment]
I think the issue is crucial. For environments / folders with lots of links, the usability of VS Code is below average.

**Comment by bpasero** [comment]
Please see #167999 for our plan this month to clean up and "groom" all opened issues. This issue was closed as part of that initiative.

**Comment by ipkiss42** [comment]
I also think this issue is crucial. Let me explain why.

Even though not expanding symlinks to a canonical path may look benign, **the impact is much bigger than it seems**. In fact, this symlink issue is at the core of multiple bugs with very different symptoms, in VS Code itself as well as in multiple extensions. Here are just a few examples:
* Tests not auto-detected:
    * https://github.com/microsoft/vscode-java-test/issues/576
    * https://github.com/microsoft/vscode-java-test/issues/630
    * https://github.com/microsoft/vscode-python/issues/18952
* Debugging-related issues:
    * https://github.com/microsoft/vscode/issues/57954
    * https://github.com/microsoft/vscode/issues/78443
    * https://github.com/microsoft/vscode/issues/117211
    * https://github.com/microsoft/vscode/issues/122924
    * https://github.com/microsoft/vscode/issues/34627
* Git-related issues:
    * https://github.com/microsoft/vscode/issues/74104#issuecomment-646276435
    * https://github.com/microsoft/vscode/issues/90611
    * https://github.com/microsoft/vscode/issues/116064
* Single file open in multiple tabs:
    * https://github.com/microsoft/vscode/issues/92679
    * https://github.com/microsoft/vscode/issues/100533
    * The current issue :)
* Miscellaneous:
    * https://github.com/redhat-developer/vscode-java/issues/2326
    * https://github.com/REditorSupport/vscode-R/issues/513
    * https://github.com/microsoft/vscode/issues/43071

The variety of these symptoms "dilutes" the votes from affected users, because most of the time users didn't get directed to a central tracking issue. The same applies to all the other issues I haven't mentioned above. And to duplicates of those issues. And to symlink-related issues whose root cause was never identified. This amounts to a large user impact, unfortunately very hard to track by votes alone due to the nature of the issue.

IIUC, resolving symlinks to their canonical path was the previous behavior of VS Code. I don't think this decision should be reverted, because it unblocked multiple use cases (see e.g. #18837), which shouldn't be broken again.

Expecting the entire ecosystem to fix these issues might work in the very long-term, but if issues keep getting resolved as "out-of scope" (like microsoft/vscode-java-test#630 or  #74104), this won't be happening anytime soon. In the meantime, users have to pay attention to always using canonical paths when opening files and folders... assuming they are even aware of the issue.

I like @koenlek's proposal in this issue for a few reasons:
* It is opt-in, so by default it won't reintroduce the issues (e.g. #18837).
* It is not opinionated regarding what an ideal long-term solution might be.
* It makes troubleshooting easier ("does the problem go away if you set `openFilesAtCanonicalPath` to `true`?").
* It allows avoiding the issue for many users at once. At my workplace, the default environment setup uses multiple symbolic links, so every developer can be impacted by the side-effects of this issue. A setting would allow for central remediation (e.g. at install time).
* It sounds easy to implement (just guessing here, I haven't looked at the actual code).

I sincerely hope you will reconsider the decision to close this issue. But regardless of what you decide, I also want to thank you for all the amazing work on VS Code!

**Comment by dewilcox** [comment]
I think this is a crucial issue too. I was going to explain my reasons, but @ipkiss42 said it better.

**Comment by bpasero** [comment]
@ipkiss42 that is a great summary, thanks! However, here the ask is to follow symlinks of files that get opened, though you list issues about opening a folder that is a symlink. If I understand the original ask here then it is to resolve symlinks when opening an editor, i.e. always check if it is a link or not. But this issue is not about resolving the possible symlink when opening a folder?

**Comment by ipkiss42** [comment]
It should apply to both scenarios. And also to opening a workspace file, if this is not covered by the regular way of opening a file.

My mental model is that all these issues, at their core, boil down to a file path comparison done somewhere. For example, the logic to decide between reusing a tab vs. creating a new one checks if there is already a tab open for the given path. If not, it opens a new one. When faced with two representations of the same path (e.g. a canonical one and one with symlinks), the comparison treats them as different and we end up with two tabs.

You can easily reproduce it:
* Open a folder (Ctrl-K Ctrl-O) via a non-canonical path.
* Open a file in that folder, via the explorer.
* Open the same file via its canonical path in the UI (Ctrl-O).
* You now have two different tabs for the same file. You can hover over each tab to see the corresponding absolute paths.

In this example, the path of the first file was constructed from the non-canonical folder path and from the relative path to the root, so it was different from the canonical path of the second file.

Conclusion: the setting is effective only if it applies to every "entry point" from where we get absolute paths.

This brings an interesting question: why do we end up with different representations of file paths today, since the UI never expands them to canonical paths? Shouldn't they all be non-canonical, thus avoiding any issue when comparing paths?
Obviously, some places in the code are making paths canonical, or they are introducing canonical paths themselves (behaving as additional "entry points"). Language servers, linters, and more generally all the tools integrating with VS Code and its extensions may provide paths which end up being compared.

Unfortunately, there is no reason why tools should all agree on using canonical or non-canonical paths. As a result:
* For tools emitting relative paths, canonical paths, or paths constructed from other paths provided by VS Code, the new setting should work just fine.
* For tools emitting non-canonical absolute paths, the new setting may introduce new problems (with similar symptoms as today).

I believe this is still better than the current situation, because:
* Tools are much more likely to go from a non-canonical path to a canonical one (a simple API call in most languages) than the opposite. IMHO, this is why the issue is so widespread today.
* VS Code (and extensions) could canonicalize paths coming from tools when the setting is active. Or the tools themselves could respect the setting.

So instead of having tools deciding independently whether or not to canonicalize paths, the setting provides a path forward for eventually aligning every tool in the ecosystem, either natively (in the tool itself) or in the integration layer.

**Comment by ipkiss42** [comment]
BTW, another way to look at it is that the setting would buy us time to figure out a better approach.

Maybe paths should be made canonical in all the places where path comparisons are done? I don't know, but it may take time to decide and act on it. I'd rather be unblocked via the setting in the meantime.

**Comment by bpasero** [comment]
Just a quick reply without going deep (yet): VS Code internally uses and identifies everything via [`URI`](https://github.com/microsoft/vscode/blob/422b581e3802e30cfb780c21d9cc1a2cd0c9f0aa/src/vs/base/common/uri.ts#L98) class which is essentially a path with a scheme. We have a canonical form of it which is simply whoever first introduces a `URI` it becomes the canonical form. This helped us avoid issues on case insensitive platforms such as macOS and Windows to not open the same file with different casing. 

However, things are more complicate than that: not only can a file be expressed by different cases and still point to the same file on disk, symbolic links can as well. However, symbolic links cannot be resolved synchronous, you need to ask the OS for its resolved form, which is a long running call. All our `URI` handling however requires synchronous execution.

Bottom line: it would probably be relatively easy to resolve a symbolic link when it comes to opening a folder to avoid opening the symbolic link but the resolved target and I think a setting for that could be added. However, resolving a link as part of opening an editor will be more tricky because essentially this requires us to make canonical `URI` handling asynchronous.

**Comment by ipkiss42** [comment]
IIUC, it wouldn't be necessary to do it when opening an editor, if the editor is opened via the explorer or via any other mechanism which relies on the (already resolved) folder path. It would be needed only when opening a file "from nothing", e.g. from the UI, in which case doing it synchronously might be acceptable.

**Comment by bpasero** [comment]
I would argue that in 99% of all cases the editor is opened from a component or piece of code that does NOT resolve the symbolic link if there is one. Not to speak about extensions who can also open any `URI` as editor where we do not know if the file is resolved or not.

**Comment by bpasero** [comment]
> My mental model is that all these issues, at their core, boil down to a file path comparison done somewhere. 
> Shouldn't they all be non-canonical, thus avoiding any issue when comparing paths?
> This brings an interesting question: why do we end up with different representations of file paths today, since the UI never expands them to canonical paths?

Yes, the issue is comparing paths, or in our case `Uri`. In many places where we deal with identity of a `Uri` we use a class called `ResourceMap`:

https://github.com/microsoft/vscode/blob/d1f42f4857c638aa1580f0ea6d0855618f7d47ad/src/vs/base/common/map.ts#L44

When deciding if a `Uri` already exists in a component, we check this map which in the end will do a path and scheme comparison: 

https://github.com/microsoft/vscode/blob/d1f42f4857c638aa1580f0ea6d0855618f7d47ad/src/vs/base/common/map.ts#L46

For editors to decide if an editor is already opened we do something similar:

https://github.com/microsoft/vscode/blob/d1f42f4857c638aa1580f0ea6d0855618f7d47ad/src/vs/workbench/contrib/files/browser/editors/fileEditorInput.ts#L429

The key thing here is that when a `Uri` enters our system, we try to use its canonical form:
* if its the first time we see a `Uri` we just use that form
* if the `Uri` exists with different casing but same value, we use the form that was used first (except for Linux)

We cannot make any assumptions about the form of the `Uri` when it enters VS Code: it can come from extensions, it can be the user clicking a file in the explorer, it can be any code that constructs a `Uri` from an existing path or by path concatenation.

In order to support symbolic links, we essentially need to resolve the symbolic link and then use the resolved value as canonical form, not the symbolic link one. The issues at hand are:
* this is a long running, relatively slow file system operation
* we currently do not allow link resolution in our file system providers (https://github.com/microsoft/vscode/issues/71204) so it would only work for local files but not in remote scenarios

Happy to discuss this further. Again, I think resolving a link when opening VS Code would be straight forward but it does not seem to address all the issues.

**Comment by dschwen** [comment]
This issue is the bane of my existence. I work in a large codebase that - as part of its build system - symlinks all headers into a single path, to keep the include path list managable. Compile errors in the terminal list the symlinks as the error location, so when clicking those errors the editor opens the symlink. **However** the actual file is likely to be open _as well_ due to teh header/source switching plugin we use. Both symlink and file open leads to frequent **corruption** of header files when enabling _format on save_ (C++). These errors are sometimes not obvious and if not caught right away will mess up your day when you continue editing the files.

**Comment by kamil-csco** [comment]
This is a frequent issue during remote debug on servers with environments that use symlinks e.g. to streamline paths. 

Fyi, PyCharm's approach was to allow the user to define path mappings:
https://www.jetbrains.com/help/pycharm/edit-project-path-mappings-dialog.html

Note, the case there is different: they work with local and remote copies of the code, while VSCode works only with a remote copy. However, using mapping would allow the user to configure with prefix path differences to "ignore" and not open a separate copy of the file (e.g., during debug).

For example: /home/user/code could be mapped to underlying path /mounts/<deep_and_complex_path_structure>/user/code avoiding the issues.

**Comment by gedalia** [comment]
It looks like part of this (Bazel sandbox python files open into a window) was turned into a issue here: https://github.com/microsoft/vscode-python/issues/17002 we (Boston Dynamics) are using debugpy and it really gets annoying that this doesn't work better, its there any way to get this improved, @brettcannon I'm happy to get you 5 upvotes if that's what it takes.

**Comment by brettcannon** [comment]
@gedalia voting is different per repo (if they use voting at all). But if there's a specific Python issue you can open it on the Python extension.

**Comment by gedalia** [comment]
@brettcannon it looks like you opened it and then closed it. Should I just create a new request with the same contents? or is this a issue with https://github.com/microsoft/debugpy hard for me to tell.

**Comment by gedalia** [comment]
looks like it might have been done in: https://github.com/microsoft/debugpy/issues/743 I'll have to test that.

**Comment by brettcannon** [comment]
@gedalia if the debugpy fix doesn't work, open a new issue on the Python debugger repo.

**Comment by StephenApptronik** [comment]
I think the issues that @bpasero raises are fairly straightforward to resolve: there are files listed in the project tree of VSCode. You only need to resolve anything when the file is opened. And when resolving, if a symlink path resolves to any of those paths, use the path that is in the project tree when opening. You can check all of fs.readlink, fs.realpath, fs.native.realpath, or any other options -- if they indicate that a file is actually one that is in the project tree, prefer to open the project tree path.

I am another developer stumbling on this issue. It's a constant problem having the wrong instance of the file open, and it is getting worse as our codebase and developer team expands.

**Comment by StephenApptronik** [comment]
At the very least, please give me a right-click on a tab that says "reopen at real path".

**Comment by BaconPancakes** [comment]
> At the very least, please give me a right-click on a tab that says "reopen at real path".

@StephenApptronik We have a workaround where we bind a keybinding into a VSCode task that resolves the symlink and opens it. Maybe it'd be useful to your team.

`tasks.json` (or `*.code-workspace`)
```
{
                "label": "Resolve Symlink",
                "type": "shell",
                "command": "code $(realpath ${file}) -r",
                "options": {
                    "shell": {
                        "args": [
                            "-c"
                        ]
                    }
                },
                "problemMatcher": [],
                "presentation": {
                    "reveal": "never",
                    "focus": false,
                    "panel": "dedicated",
                    "showReuseMessage": false,
                    "close": true
                }
            }, 
```

`keybindings.json`
```
[
    {
        "key": "f3",
        "command": "workbench.action.tasks.runTask",
        "args": "Resolve Symlink",
        "when": "editorFocus"
    },
]
```

**Comment by wackywendell** [comment]
Linking issues - the option suggested here appears to be a solution to the issue in #100533. Note that there was some interest on that issue as well - 26 👍 , 40ish comments - which I think would be interest in the solution here!

Edit: Just noticed this was mentioned above in https://github.com/microsoft/vscode/issues/130082#issuecomment-1342773594; I missed that on my first quick read-through.
